src/ucp/core/ucp_am.c

```diff
@@ -470,43 +479,86 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
     return status;
 }
 
-static ucs_status_t ucp_am_zcopy_single(uct_pending_req_t *self)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_am_zcopy_single_common(uct_pending_req_t *self, ucp_am_hdr_t *hdr,
+                           size_t am_hdr_length, unsigned user_hdr_length,
+                           uint8_t am_id)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
-    ucp_am_hdr_t hdr;
+    size_t hdr_length  = am_hdr_length + user_hdr_length;
+    ucs_status_t status;
 
-    ucp_am_fill_header(&hdr, req);
+    ucp_am_fill_header(hdr, req);
+    ucp_am_pack_user_header(UCS_PTR_BYTE_OFFSET(hdr, am_hdr_length), req);
 
-    return ucp_do_am_zcopy_single(self, UCP_AM_ID_SINGLE, &hdr,
-                                  sizeof(hdr), ucp_proto_am_zcopy_req_complete);
+    req->send.length -= user_hdr_length;
+    status            = ucp_do_am_zcopy_single(self, am_id, hdr, hdr_length,
+                                               ucp_proto_am_zcopy_req_complete);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        req->send.length += user_hdr_length;
+    }
+
+    return status;
 }
 
-static ucs_status_t ucp_am_zcopy_single_reply(uct_pending_req_t *self)
+static ucs_status_t ucp_am_zcopy_single(uct_pending_req_t *self)
 {
-    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
-    ucp_am_reply_hdr_t reply_hdr;
+    ucp_request_t *req       = ucs_container_of(self, ucp_request_t, send.uct);
+    unsigned user_hdr_length = req->send.msg_proto.am.header_length;
+    size_t hdr_length        = sizeof(ucp_am_hdr_t) + user_hdr_length;
+    ucp_am_hdr_t *hdr        = ucs_alloca(hdr_length);
 
-    ucp_am_fill_header(&reply_hdr.super, req);
-    reply_hdr.ep_ptr = ucp_request_get_dest_ep_ptr(req);
+    return ucp_am_zcopy_single_common(self, hdr, sizeof(*hdr), user_hdr_length,
+                                      UCP_AM_ID_SINGLE);
+}
 
-    return ucp_do_am_zcopy_single(self, UCP_AM_ID_SINGLE_REPLY,
-                                  &reply_hdr, sizeof(reply_hdr),
-                                  ucp_proto_am_zcopy_req_complete);
+static ucs_status_t ucp_am_zcopy_single_reply(uct_pending_req_t *self)
+{
+    ucp_request_t *req            = ucs_container_of(self, ucp_request_t,
+                                                     send.uct);
+    unsigned user_hdr_length      = req->send.msg_proto.am.header_length;
+    size_t hdr_length             = sizeof(ucp_am_reply_hdr_t) +
+                                    user_hdr_length;
+    ucp_am_reply_hdr_t *reply_hdr = ucs_alloca(hdr_length);
```

Comment:
can user hdr be very big? what limits the stack allocation?

Reply:
thought about it..
With zcopy, user header is limited by maximal UCT AM header (which may be quite large for certain transports)
do you think better use calloc starting from certain size?

Reply:
IMO we should have zcopy for user header as well, by using bounce buffer mpool